### PR TITLE
Singly overlappy types

### LINKED ISSUE #28751: Overlappy type checking not working when both types have only one property
```ts
interface CatDog { cat, dog }
interface ManBearPig { man, bear, pig }
interface Platypus { platypus }

type ExoticAnimal =
    | CatDog
    | ManBearPig
    | Platypus;

declare function addToZoo(animal: ExoticAnimal): void;

addToZoo({ dog: "Barky McBarkface" });
```

This is because the call to `getIntersectionType` will produce a single intersection type (`"platypus" & "dog"`) instead of `never`. Note that the following *will* work:

```ts
interface CatDog { cat, dog }
interface ManBearPig { man, bear, pig }
interface Platypus { platy, pus }

type ExoticAnimal =
    | CatDog
    | ManBearPig
    | Platypus;

declare function addToZoo(animal: ExoticAnimal): void;

let x = { dog: "Barky McBarkface"};
addToZoo(x)
```

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
With this PR, overlap checking is only performed if overlap consists of a single index query (`keyof T`) or of purely literal types.

Fixes #28751.

~Also partially subsumes the intent of #28695 even though this discounts types who both have something like a `number` index signature. So whereas we used to give improved errors for something like `Array<string>` to `ConcatArray<number> | { cat: "meow" }`, this PR causes a regression on there.~

### CODE REVIEW

**Comment by weswigham** [comment]
Doesn't just forcing the collapse of empty intersections with a union wrapper solve the issue without causing a regression in the comparison of subtypes with index signatures?

**Comment by DanielRosenwasser** [comment]
I sent a change to just go back to the original proposal - only count single unit types.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -11986,7 +11986,7 @@ namespace ts {
                             matchingCount = len;
                         }
                     }
-                    else if (!(overlap.flags & TypeFlags.Never) && 1 >= matchingCount) {
+                    else if (isUnitType(overlap) && 1 >= matchingCount) {
```
Comment (weswigham) [inline src/compiler/checker.ts]: `overlap.flags & TypeFlags.DisjointDomains` is probably better than straight `isUnitType` - this way we also capture `number` and `string` index signature overlap.

Comment (DanielRosenwasser) [reply src/compiler/checker.ts]: That would weigh a numeric indexer equally with a common property, which isn't ideal.

**Pull request merged**